### ISSUE TITLE
Use automatic differentiation

### ISSUE THREAD

**Issue by dpsanders** [WyLyu/UPOsHam#37]
I suggest using automatic differentiation to avoid having to manually specify derivative functions.

**Comment by Shibabrat** [comment]
@kthyng @dpsanders 
We are happy to do this for a later version of this package as this suggested feature will certainly improve the usability.

**Comment by kthyng** [comment]
@dpsanders @Shibabrat I am not familiar with this but it sounds like it is probably beyond the scope of improving the basic code structure and setup of this package. Agreed that it is a helpful suggestion for future development.

@dpsanders please let me know if you can agree with this.

**Comment by dpsanders** [comment]
Yes I agree that it is beyond the basic restructuring requirement, but would be a huge usability plus for a future version.

**Comment by kthyng** [comment]
@dpsanders You are doing a huge service for @Shibabrat by giving this in-depth feedback. However, it is beyond the scope for the purposes of this software as a JOSS submission and the review may proceed.
